File: packages/android/src/main/java/com/markdown2ui/FormatUtils.kt
package com.markdown2ui

import java.text.NumberFormat
import java.util.Currency
import java.util.Locale

/** Format a number as a whole integer string when it has no fractional part. */
private fun formatWholeNumber(value: Number): String {
    val d = value.toDouble()
    return if (d % 1 == 0.0) d.toLong().toString() else value.toString()
}

fun formatDisplayValue(
    value: Number,
    format: FormatAnnotation?,
    locale: Locale = Locale.getDefault(),
): String {
    if (format == null) return formatWholeNumber(value)

    return when (format) {
        is FormatAnnotation.Currency -> {
            val nf = NumberFormat.getCurrencyInstance(locale)
            try {
                nf.currency = Currency.getInstance(format.code)
            } catch (_: IllegalArgumentException) {
                // Fall back to default currency if code is invalid
            }
            nf.format(value)
        }

        is FormatAnnotation.Unit -> {
            val unitLabel = if (value.toDouble() != 1.0 && format.plural != null) {
                format.plural
            } else {
                format.unit
            }
            val formatted = formatWholeNumber(value)
            "$formatted $unitLabel"
        }

        is FormatAnnotation.Percent -> {
            val nf = NumberFormat.getPercentInstance(locale)
            nf.format(value.toDouble() / 100.0)
        }

        is FormatAnnotation.Integer -> {
            val nf = NumberFormat.getIntegerInstance(locale)
            nf.format(value.toLong())
        }

        is FormatAnnotation.Decimal -> {
            val nf = NumberFormat.getNumberInstance(locale)
            nf.minimumFractionDigits = format.places
            nf.maximumFractionDigits = format.places
            nf.format(value)
        }
    }
}

File: packages/ios/Sources/Markdown2UI/FormatUtils.swift
import Foundation

public func formatDisplayValue(_ value: Double, format: FormatAnnotation?, locale: Locale = .current) -> String {
    guard let format = format else {
        return NumberFormatter.localizedString(from: NSNumber(value: value), number: .decimal)
    }

    switch format {
    case .currency(let code):
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = code
        formatter.locale = locale
        return formatter.string(from: NSNumber(value: value)) ?? "\(value)"

    case .unit(let unit, let plural):
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = locale
        let numStr = formatter.string(from: NSNumber(value: value)) ?? "\(value)"
        let unitLabel: String
        if value == 1 {
            unitLabel = unit
        } else {
            unitLabel = plural ?? "\(unit)s"
        }
        return "\(numStr) \(unitLabel)"

    case .percent:
        let formatter = NumberFormatter()
        formatter.numberStyle = .percent
        formatter.locale = locale
        formatter.multiplier = 1
        return formatter.string(from: NSNumber(value: value)) ?? "\(value)%"

    case .integer:
        let formatter = NumberFormatter()
        formatter.numberStyle = .none
        formatter.locale = locale
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? "\(Int(value))"

    case .decimal(let places):
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = locale
        formatter.minimumFractionDigits = places
        formatter.maximumFractionDigits = places
        return formatter.string(from: NSNumber(value: value)) ?? String(format: "%.\(places)f", value)
    }
}
